src/v/kafka/server/group_tx_tracker_stm.cc

```diff
@@ -0,0 +1,181 @@
+/*
+ * Copyright 2024 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#include "kafka/server/group_tx_tracker_stm.h"
+
+namespace kafka {
+
+group_tx_tracker_stm::group_tx_tracker_stm(
+  ss::logger& logger, raft::consensus* raft)
+  : raft::persisted_stm<>("group_tx_tracker_stm.snapshot", logger, raft)
+  , group_data_parser<group_tx_tracker_stm>() {}
+
+void group_tx_tracker_stm::maybe_add_tx_begin_offset(
+  kafka::group_id group, model::producer_identity pid, model::offset offset) {
+    auto [it, inserted] = _all_txs.try_emplace(group, pid, offset);
+    if (!inserted) {
+        // group already exists
+        it->second.maybe_add_tx_begin(pid, offset);
+    }
+}
+
+void group_tx_tracker_stm::maybe_end_tx(
+  kafka::group_id group, model::producer_identity pid) {
+    auto it = _all_txs.find(group);
+    if (it == _all_txs.end()) {
+        return;
+    }
+    auto& group_data = it->second;
+    auto p_it = group_data.producer_to_begin.find(pid);
+    if (p_it == group_data.producer_to_begin.end()) {
+        return;
+    }
+    group_data.begin_offsets.erase(p_it->second);
+    group_data.producer_to_begin.erase(p_it);
+    if (group_data.producer_to_begin.empty()) {
+        _all_txs.erase(group);
+    }
+}
+
+ss::future<> group_tx_tracker_stm::apply(const model::record_batch& b) {
+    auto holder = _gate.hold();
+    co_await parse(b.copy());
+}
+
+model::offset group_tx_tracker_stm::max_collectible_offset() {
+    auto result = last_applied_offset();
+    for (const auto& [_, group_state] : _all_txs) {
```

Comment:
hmm I wonder what the scale here is? Maybe we should just recalculate it once in maybe_end_tx?

Reply:
few 100s to max 1000 at the higher end I can think of. Also do note this inflight state, that is groups with open transaction (not historical groups in the log), so very unlikely this is in double digits IMO. Adding to that this is only called from compaction patch attempting to compute the compaction boundary, which is not a hot path.

Reply:
ok, sounds good